### PR TITLE
Revisit long members in a type declaration.

### LINKED ISSUE #1362: Aesthetics of long members in a type declaration
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAkgGKyxgC8AOgHYQQCGARgM7oBOd06YAQgDZ0ALMGgBmJYWBasAllQDmYALQA%2BMAzqsJU2QpVq6ALy1sdS1bPSQANOChNE6dDqaowAbVsRQtWpAAkWLiukBbWnr5%2BAG50vACuyGgePj4ADDbJvvhUSFToAMrSBsjpGWAALOEQALrhAL4lXpVQATgJUKENEdFxbUmlAIydPpAAsnQAHgAysoiTiPLoAmGlYABsKU01PvXh3sn%2BgW2QDCS8y-tRMfGufRkATEO%2BeYgAttJwvLBUAILoAKLZADyImmVGKTQgIhi9k2dU6e2GLSCaGOp3OiO610SELAAGZHhBIHlsBxENxEGJWIgAAoaOgvByIVjo5JseKw7bwpoHVrBE6wM6PfyY3o4soEqDE0nkymzWAAdyZ0Do9iykVgyqcXxZPjZiA5tB2PgREUOfLRQsuPRuOIArBKiSToGSKbAqQBVbC4VjK1VUdWa6TaiVQ3gwjJbQ1cjI85FQfmC7lWrHuHGrB1S50yt2IADC-CYTFzXyksU4bp1tFD4eSkYgRtoJsJSKOCcrwquopWAHYM06XbKRq8GEzK5DofqI3DdkmzSi25aRTaVgAOPvS11U4ufKhjsDVye16fG2e8%2BcWpNL7ErACc6%2Bd3xE6CZxZeLzoe71Bvr0Yuc-jF4xsmXYDGkOKOhusrPG8HzBjiB7fmADaNEB-6ogK7bAcuAyDOBmaII%2Bz6sNB7wCnBKxflOnIzqhZ4ARhi6dthGT9A84FZDk6CAlQAAqrCYAA6tIix7ghVFRjRf50ehiZAVeqYrP0%2BJ4f23ysLAsTZAAIogvDSG8RGfqw7LiT%2BkmImhHSXkx14DOK4FjOM%2BAiDxAjzH8YaIHkAhuugQnwCJEplBsplIb%2BFnSVZck2QpAz2g5ExZCI0jjIC3p0Ogbp-OM2BUoWQa7hKtohUe1EnrRcYhLkmHybcyT9OmCXjAASogcCsPA-mBWKJU%2BHWYXmaakXVYx1q2SxvZNa17XwAAcrELwjqwwL4M%2BLwuBK-SIchYBNs0aEvLEvBOHpYLEKw76OEyPG8qNKZ1T4-RruB01uvAIxHSdMznZdhkOtAAh0pwTIAPrytIAVLNt4VDZVUUXLVOL9HeTVqewmCAqw0wsF1SxBb1tD9Tte2xkc8MYjFD20HcYErKMExo3QGNY9ILDzYtTIrWtG1I9Dg3Ngdn3SKdiA-RlRE3UEd0gfcuF04zzPY%2BgH3HcL31ur9o7-YD7DA6wYMQyJfPlVJcMjdZY2xfcbF045ABqnbcLI8A6Lje7BcbjanmbljS8xyR3MptsTIQmmcAVTvZK7ht4z1nsoabZPm9FltUxAdz2cH4xabA6AAOIONluWIPlXxu0VBPVMeXsVa2gEI5TOJ3PFttCyLfAagA1twuudw4TDcXk9J5gKC2FfBE7x7t3t1wxFv3U3jV07NiDyu3DiKvMkuIDpyVUMJBXfNkQ4c6wPMrGJpUSSbEWVQu88ywHk10wA0ogiDYM5rnzFkQ8MqC4IL6T1CsTGewRyZdEbisO4z0s4eTeEwAQ5c44gJhgLaS98U4L2gSjOmBR5C8AIqwOQC15h%2BUQAwEYCBAGlEvn1auCdb6z1kg3VOOJcS01KJAb4ek5BUFDlQcOXwCh8IyrEKkPFYAcTIRlAqolgFXzMjfWGzCapQNKLiOWXCeFESoLIyIcp5AnyWrvWQB8SzyM8lPEmLZzRzywY-HwuIbZcJVk4ABkx6QMHgHQfMsAmA6BXmvGYliaz0LKjXROdiWEUzYSsXEQcuFaVZowQh8DWYCDyJgXIExQmHnCdfSJTDgjzHgMCABeQsCELUXEjRmcuEAjKSCEJ-1WC8BEJAaxYDzz2NYdgjRLcuGVJkJwKhSA8nbVsFUEoIQmCEACa4BCtQgA)

#### Code

```fsharp
type IFoo =
    abstract Blah : foo : string -> bar : string -> baz : string -> int
```
            

#### Result

```fsharp
type IFoo =
    abstract Blah: foo:string
     -> bar:string -> baz:string -> int

```
            
#### Problem description

I set the line length very short so as to trigger this.

This is purely stylistic, but I would prefer to follow a different rule: if a member declaration has to be split over multiple lines, then all the arguments should be split onto their own line.

```fsharp
type IFoo =
    abstract Blah :
        foo : string ->
        bar : string ->
        baz : string ->
        int
```

I recognise that this change should be accompanied by a change to the G-Research style guidelines, which are currently silent on the matter. If you think this is a reasonable choice, I'll raise an appropriate PR to https://github.com/G-Research/fsharp-formatting-conventions/ .

#### Extra information

- [ ] The formatted result breaks by code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas Master at 01/09/2021 08:58:25 - ca42f0d7cf05144c5838bba725378e2a856211bb

```fsharp
    { config with
                MaxLineLength = 60 }
```

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#ignore-files-fantomasignore)?</sub>

### PR DESCRIPTION
**Pull request by nojaf**
Fixes #1362.
@Smaug123 having the arrows at the start or at the end is not really a big deal for Fantomas.

### CODE REVIEW

File: src/Fantomas/CodePrinter.fs [hunk 1]
```diff
@@ -4005,67 +4003,60 @@ and genPrefixTypes astContext node ctx =
...
 and genTypeList astContext node =
-    match node with
-    | [] -> sepNone
-    | (t, [ ArgInfo (ats, so, isOpt) ]) :: ts ->
-        let gt =
+    let gt (t, args: SynArgInfo list) =
+        match t, args with
+        | TTuple ts', _ ->
+            // The '/' separator shouldn't appear here
+            let hasBracket = not node.IsEmpty
```
Comment (Smaug123) [inline src/Fantomas/CodePrinter.fs:4010]: I don't understand what you mean by "the '/' separator". Could you explain that bit please?

**Review by Smaug123 — APPROVED** [review]

File: src/Fantomas/CodePrinter.fs (hunk 1 above)
Comment (nojaf) [reply src/Fantomas/CodePrinter.fs:4010]: To be honest, I kind of found that comment when I started.
https://github.com/fsprojects/fantomas/blob/924f63899a692c58d9bf4eb014d071f10d7e5e96/src/Fantomas/CodePrinter.fs#L4031

I can't really say what it does either. Perhaps I could remove the comment.

**Review by Smaug123 — COMMENTED** [review]
Actually, could we please have a test case for when the two issues happen together?

```fsharp
type IFoo =
    abstract Bar :
        i : int
        -> a : string
        * foo : int
        -> string
```
or whatever happens? (As long as something reasonably sane happens, I don't mind what it is.)

**Review by Smaug123 — APPROVED** [review]
Happy with the additional tests, thanks!

**Pull request merged**